core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -192,13 +199,49 @@ static boolean execute(SearchContext searchContext, final IndexSearcher searcher
 
             boolean timeoutSet = scrollContext == null && searchContext.timeout() != null &&
                 searchContext.timeout().equals(SearchService.NO_TIMEOUT) == false;
+
+            final Runnable timeoutRunnable;
             if (timeoutSet) {
-                // TODO: change to use our own counter that uses the scheduler in ThreadPool
-                // throws TimeLimitingCollector.TimeExceededException when timeout has reached
-                collectors.add(createTimeoutCollectorContext(searchContext.timeEstimateCounter(), searchContext.timeout().millis()));
+                final Counter counter = searchContext.timeEstimateCounter();
+                final long startTime = counter.get();
+                final long timeout = searchContext.timeout().millis();
+                final long maxTime = startTime + timeout;
+                timeoutRunnable = () -> {
+                    final long time = counter.get();
+                    if (time > maxTime) {
+                        queryResult.searchTimedOut(true);
+                        throw new CollectionTerminatedException();
```

Comment:
You could use a TimeExceededException to stop the collection on all segments ? Otherwise you also need to check the timeout when the leafCollector is created like the CancellableCollector does ? If you prefer the second option you can remove the try/catch around the searcher.search below since this code does not throw TimeExceededException anymore.